pkg/controller/podupgrade/pod_upgrade_controller.go

```diff
@@ -0,0 +1,340 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podupgrade
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	appsinformers "k8s.io/client-go/informers/apps/v1"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	client "k8s.io/client-go/kubernetes"
+	appslisters "k8s.io/client-go/listers/apps/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+type Controller struct {
+	kubeclientset client.Interface
+
+	daemonsetLister appslisters.DaemonSetLister
+	daemonsetSynced cache.InformerSynced
+	nodeLister      corelisters.NodeLister
+	nodeSynced      cache.InformerSynced
+	podLister       corelisters.PodLister
+	podSynced       cache.InformerSynced
+
+	daemonsetWorkqueue workqueue.RateLimitingInterface
+	nodeWorkqueue      workqueue.Interface
+}
+
+func NewController(kc client.Interface, daemonsetInformer appsinformers.DaemonSetInformer,
+	nodeInformer coreinformers.NodeInformer, podInformer coreinformers.PodInformer) *Controller {
+	// TODO: Is eventBroadCaster needed?
+	ctrl := Controller{
+		kubeclientset:   kc,
+		daemonsetLister: daemonsetInformer.Lister(),
+		daemonsetSynced: daemonsetInformer.Informer().HasSynced,
+
+		nodeLister: nodeInformer.Lister(),
+		nodeSynced: nodeInformer.Informer().HasSynced,
+
+		podLister: podInformer.Lister(),
+		podSynced: podInformer.Informer().HasSynced,
+
+		daemonsetWorkqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		nodeWorkqueue:      workqueue.New(),
+	}
+
+	daemonsetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, new interface{}) {
+			newDS := new.(*appsv1.DaemonSet)
+			oldDS := old.(*appsv1.DaemonSet)
+
+			// Only control daemonset with annotation "apps.openyurt.io/upgrade-strategy"
+			if !metav1.HasAnnotation(newDS.ObjectMeta, UpgradeAnnotation) || !metav1.HasAnnotation(oldDS.ObjectMeta, UpgradeAnnotation) {
+				return
+			}
+
+			// TODO: change to compare generation and revision hash
+			if newDS.ResourceVersion == oldDS.ResourceVersion || reflect.DeepEqual(newDS.Spec.Template.Spec, oldDS.Spec.Template.Spec) {
+				return
+			}
+
+			ctrl.enqueueDaemonset(new)
+		},
+	})
+
+	nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, new interface{}) {
+			oldNode := old.(*corev1.Node)
+			newNode := new.(*corev1.Node)
+			if !NodeReady(&oldNode.Status) && NodeReady(&newNode.Status) {
+				klog.Infof("node %q turn to ready", newNode.Name)
+				ctrl.nodeWorkqueue.Add(newNode.Name)
```

Comment:
please calculate the corresponding OTA and auto daemonsets, and queue these daemonsets into daemonset workqueue.